### PR TITLE
Threading - TLS connections on multiple threads

### PR DESCRIPTION
**Pull request by ashamis**
Process TLS connections on multiple threads

part of #774

### CODE REVIEW

**Comment by cimetrics** [comment]
threading_alloc_free_8@[4476 aka 20200202.2](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=4476&view=results) vs master ewma over 30 builds from [4057](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=4057&view=results) to [4472](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=4472&view=results)
images

File: src/ds/thread_messaging.h [hunk 1]
```diff
@@ -18,11 +18,20 @@ extern std::map<std::thread::id, uint16_t> thread_ids;
 
 namespace enclave
 {
+  const uint64_t magic_const = 0xba55ball;
```
Comment (achamayou) [inline src/ds/thread_messaging.h]: Why do we have a magic const?

File: src/enclave/tlsendpoint.h [hunk 2]
```diff
@@ -130,6 +142,10 @@ namespace enclave
           return data;
       }
 
+      if (ctx == nullptr)
+      {
+        throw std::exception();
```
Comment (jumaffre) [inline src/enclave/tlsendpoint.h]: These are rather generic exceptions?

File: src/enclave/tlsendpoint.h [hunk 3]
```diff
@@ -130,6 +142,10 @@ namespace enclave
           return data;
       }
 
+      if (ctx == nullptr)
```
Comment (jumaffre) [inline src/enclave/tlsendpoint.h]: How could this condition be true?

File: src/enclave/tlsendpoint.h [hunk 4]
```diff
@@ -195,12 +211,42 @@ namespace enclave
...
     {
+      if (thread_ids[std::this_thread::get_id()] != execution_thread)
+      {
+        throw std::exception();
+      }
       pending_read.insert(pending_read.end(), data, data + size);
       do_handshake();
     }
 
+    struct send_raw_msg
```
Comment (jumaffre) [inline src/enclave/tlsendpoint.h]: Minor: struct name should be `SendRawMsg`

File: src/enclave/httpendpoint.h [hunk 5]
```diff
@@ -95,7 +117,29 @@ namespace enclave
       session_id(session_id)
     {}
 
+    struct SendCbMsg
```
Comment (jumaffre) [inline src/enclave/httpendpoint.h]: This is the same type as `send_raw_msg` in `tlsendpoint.h`. Could we remove this one?

File: src/enclave/httpendpoint.h [hunk 6]
```diff
@@ -27,7 +27,29 @@ namespace enclave
       p(parser_type, *this)
     {}
 
+    struct RecvCbMsg
```
Comment (jumaffre) [inline src/enclave/httpendpoint.h]: Could we remove this struct as well?

File: src/enclave/httpendpoint.h [hunk 7]
```diff
@@ -133,6 +177,92 @@ namespace enclave
       flush();
     }
 
+    struct HandleProcessCbMsg
+    {
+      std::shared_ptr<Endpoint> self;
+      std::shared_ptr<JsonRpcContext> rpc_ctx;
+      std::shared_ptr<RpcHandler> search;
```
Comment (jumaffre) [inline src/enclave/httpendpoint.h]: Minor: this member should be called frontend or something similar.

File: src/enclave/httpendpoint.h [hunk 8]
```diff
@@ -133,6 +177,92 @@ namespace enclave
...
+    {
+      if (msg->data.d.size() != 0)
+      {
+        reinterpret_cast<HTTPServerEndpoint*>(msg->data.self.get())
+          ->send_response(msg->data.d);
+      }
+      else
+      {
+        reinterpret_cast<HTTPServerEndpoint*>(msg->data.self.get())
+          ->send_response(nullptr);
```
Comment (jumaffre) [inline src/enclave/httpendpoint.h]: Can you pass a `nullptr` to `send_response()`?

File: src/enclave/httpendpoint.h [hunk 9]
```diff
@@ -133,6 +177,92 @@ namespace enclave
...
+            &send_response_vect);
+          msg->data.status = HTTP_STATUS_OK;
+          msg->data.self = this->shared_from_this();
+          if (response.has_value())
+          {
+            msg->data.d = response.value();
+          }
+          else
+          {
+            msg->data.d.resize(0);
```
Comment (jumaffre) [inline src/enclave/httpendpoint.h]: If a RPC has been forwarded, I believe we can stop execution here and not return anything.

Comment (jumaffre) [reply src/enclave/httpendpoint.h]: Oh wait, we can never execute this code path because we already know that `response.has_value()` here.

**Pull request merged**